enterprise/server/backends/pebble_cache/pebble_cache.go

```diff
@@ -2401,7 +2610,139 @@ func (e *partitionEvictor) partitionKeyPrefix() string {
+		cumulativeCount += v.count
+		if cumulativeCount != v.cumulativeCount {
+			sb.WriteString(fmt.Sprintf("CUMULATIVE COUNT MISMATCH, expected %d but was %d\n", cumulativeCount, v.cumulativeCount))
+		}
+		sb.WriteString(fmt.Sprintf("%23s %20d %20d\n", v.groupID, v.count, v.cumulativeCount))
+	}
+	if cumulativeCount != e.groupIDApproxTotalCount {
+		sb.WriteString(fmt.Sprintf("TOTAL COUNT MISMATCH %d vs %d\n", cumulativeCount, e.groupIDApproxTotalCount))
+	}
+	return sb.String()
+}
+
+// samples a random group ID and updates the approximate count for that group
+func (e *partitionEvictor) updateGroupIDApproxCounts() {
+	randomGroupID, err := e.sampleGroupID()
+	if status.IsNotFoundError(err) {
+		return
+	}
+	if err != nil {
+		log.Warningf("could not sample group ID: %s", err)
+		return
+	}
+
+	approxCount, err := e.approxGroupItemCount(randomGroupID)
+	if err != nil {
+		log.Warningf("could not approximate group count for %s: %s", randomGroupID, err)
+		return
+	}
+
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	sampledGroupIdx := -1
+	for i, w := range e.groupIDApproxCounts {
+		if w.groupID == randomGroupID {
+			sampledGroupIdx = i
+		}
+	}
+
+	// If the group ID is already in the list, we need to update the count.
+	if sampledGroupIdx != -1 {
+		data := e.groupIDApproxCounts
+		oldCount := data[sampledGroupIdx].count
+		delta := approxCount - oldCount
+		data[sampledGroupIdx].count = approxCount
+		data[sampledGroupIdx].cumulativeCount += delta
+		e.groupIDApproxTotalCount += delta
+
+		// Shift the entry up or down as necessary to keep the list sorted.
+
+		for i := sampledGroupIdx - 1; i >= 0; i-- {
+			if data[i].count <= data[i+1].count {
+				break
+			}
+			tmp := data[i]
+			data[i] = data[i+1]
+			data[i].cumulativeCount -= tmp.count
+			data[i+1] = tmp
+			data[i+1].cumulativeCount += approxCount
+		}
+
+		for i := sampledGroupIdx + 1; i < len(data); i++ {
+			if data[i-1].count <= data[i].count {
+				data[i].cumulativeCount += delta
+				continue
+			}
+			tmp := data[i-1]
+			data[i-1] = data[i]
+			data[i-1].cumulativeCount -= oldCount
+			data[i] = tmp
+			data[i].cumulativeCount += data[i-1].count
+		}
+	} else {
+		e.groupIDApproxTotalCount += approxCount
+		// Not in the list, need to insert it.
+		idx, _ := slices.BinarySearchFunc(e.groupIDApproxCounts, groupIDApproxCount{count: approxCount}, func(a, b groupIDApproxCount) int {
+			if a.count < b.count {
+				return -1
+			}
+			if a.count > b.count {
+				return 1
+			}
+			return 0
+		})
+		cumulativeCount := approxCount
+		if idx > 0 {
+			cumulativeCount = e.groupIDApproxCounts[idx-1].cumulativeCount + approxCount
+		}
+		w := groupIDApproxCount{groupID: randomGroupID, count: approxCount, cumulativeCount: cumulativeCount}
+		e.groupIDApproxCounts = append(e.groupIDApproxCounts[:idx], append([]groupIDApproxCount{w}, e.groupIDApproxCounts[idx:]...)...)
+		for i := idx + 1; i < len(e.groupIDApproxCounts); i++ {
+			e.groupIDApproxCounts[i].cumulativeCount += approxCount
+		}
+	}
```

Comment:
this feels kind of error prone to me, but I cannot find fault with it.

It's the combination of for loops with offset indices (sampledGroupIdx + 1, sampledGroupIdx - 1), array indexes (data[i-1] = data[i]), temp variables and deltas being applied to things.

Some simpler version might be just removing then re-sorting the entire slice when editing it?

You know better than I whether or not the performance gains warrant this complexity, but I think personally I'd prefer starting with something simpler and only adding this if it was critical, especially if this only fires every N seconds or so.

Reply:
Went with something simpler for now. We can revisit later.